### PR TITLE
GEM-103 | added internal/password & internal/password-recovery 

### PR DESCRIPTION
**Pull request by pwgacek**

### CODE REVIEW

**Comment by pwgacek** [comment]
image

**Comment by pwgacek** [comment]
image

**Comment by pwgacek** [comment]
image

File: src/integration/kotlin/pl/edu/agh/gem/integration/BaseIntegrationSpec.kt [hunk 1]
```diff
@@ -16,6 +17,8 @@ import pl.edu.agh.gem.integration.environment.ProjectConfig
 )
 @ActiveProfiles("integration")
 abstract class BaseIntegrationSpec(body: ShouldSpec.() -> Unit) : ShouldSpec(body) {
+
+    override fun testCaseOrder(): TestCaseOrder? = TestCaseOrder.Sequential
```
Comment (lawos98) [inline src/integration/kotlin/pl/edu/agh/gem/integration/BaseIntegrationSpec.kt:21]: why we added that ? could you import that directly ?

File: src/main/kotlin/pl/edu/agh/gem/external/client/JavaMailSenderClient.kt [hunk 2]
```diff
@@ -1,23 +1,41 @@
...
         try {
-            javaMailSender.send(email)
+            val mimeMessage: MimeMessage = javaMailSender.createMimeMessage()
+            val helper = MimeMessageHelper(mimeMessage, MULTIPART_MODE_RELATED, UTF_8.name())
+            helper.setFrom(emailProperties.username)
+            helper.setTo(email)
+            helper.setSubject(subject)
+            helper.setText(text, true)
+
+            val contentId = "icon"
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/external/client/JavaMailSenderClient.kt]: could you extract that to config ? with icon etc

File: src/main/kotlin/pl/edu/agh/gem/external/client/JavaMailSenderClient.kt [hunk 3]
```diff
@@ -1,23 +1,41 @@
...
-    override fun sendEmail(email: SimpleMailMessage) {
+    override fun sendEmail(email: String, subject: String, text: String) {
         try {
-            javaMailSender.send(email)
+            val mimeMessage: MimeMessage = javaMailSender.createMimeMessage()
+            val helper = MimeMessageHelper(mimeMessage, MULTIPART_MODE_RELATED, UTF_8.name())
+            helper.setFrom(emailProperties.username)
+            helper.setTo(email)
+            helper.setSubject(subject)
+            helper.setText(text, true)
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/external/client/JavaMailSenderClient.kt]: could you create for that factory or sth like that to not create that in client

File: src/main/kotlin/pl/edu/agh/gem/internal/service/EmailService.kt [hunk 4]
```diff
@@ -1,26 +1,42 @@
...
-    private val emailProperties: EmailProperties,
+    private val fileReader: FileReader,
 
 ) {
     fun sendVerificationEmail(emailDetails: VerificationEmailDetails) {
-        val email = SimpleMailMessage()
-        email.from = emailProperties.username
-        email.setTo(emailDetails.email)
+        val subject = "Group Expense Manager - Verification code"
+        val text = fileReader.read("templates/verification.html")
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/internal/service/EmailService.kt]: could you also extract that const to some config or factory ?

Comment (lawos98) [reply src/main/kotlin/pl/edu/agh/gem/internal/service/EmailService.kt]: or companions object to have that const for username code etc

File: src/test/kotlin/pl/edu/agh/gem/internal/service/EmailServiceTest.kt [hunk 5]
```diff
@@ -5,31 +5,68 @@ import org.mockito.kotlin.anyVararg
...
+        // given
+        val passwordRecoveryEmailDetails = createPasswordRecoveryEmailDetails()
+        whenever(fileReader.read(anyVararg())).thenReturn("<html></html>")
+
+        // when
+        emailService.sendPasswordRecoveryEmail(passwordRecoveryEmailDetails)
+
+        // then
+        verify(externalEmailSenderClient, times(1)).sendEmail(
+            anyVararg(String::class),
```
Comment (lawos98) [inline src/test/kotlin/pl/edu/agh/gem/internal/service/EmailServiceTest.kt]: why we use anyVararg? if that method have type string ( not for example any() )

File: src/integration/kotlin/pl/edu/agh/gem/integration/BaseIntegrationSpec.kt (hunk 1 above)
Comment (pwgacek) [reply src/integration/kotlin/pl/edu/agh/gem/integration/BaseIntegrationSpec.kt:21]: didn't figured out how to reset mocked email server for each test so i had to determine order of tests so i can do the assertions

**Comment by pwgacek** [comment]
przesunąłem kod i hasło na środek: 
image
image

**Review by lawos98 — APPROVED** [review]

File: src/integration/kotlin/pl/edu/agh/gem/integration/BaseIntegrationSpec.kt (hunk 1 above)
Comment (lawos98) [reply src/integration/kotlin/pl/edu/agh/gem/integration/BaseIntegrationSpec.kt:21]: you could restart each time using 'AfterTestListener'  interface

File: src/main/kotlin/pl/edu/agh/gem/external/client/JavaMailSenderClient.kt (hunk 2 above)
Comment (lawos98) [reply src/main/kotlin/pl/edu/agh/gem/external/client/JavaMailSenderClient.kt]: not extracted but ok

**Pull request merged**